### PR TITLE
Add cell->face_iterators()

### PR DESCRIPTION
**Pull request by peterrum**
references #8563

ping @bangerth @masterleinad

### CODE REVIEW

**Comment by peterrum** [comment]
Please note that I still need to write a change log entry, a test, and some documentation.

File: include/deal.II/grid/tria_accessor.templates.h [hunk 1]
```diff
@@ -3173,6 +3173,22 @@ CellAccessor<dim, spacedim>::face(const unsigned int i) const
 
 
 
+template <int dim, int spacedim>
+inline std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>, 2 * dim>
```
Comment (masterleinad) [inline include/deal.II/grid/tria_accessor.templates.h]:
```suggestion
inline std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>, GeometryInfo<dim>::faces_per_cell>
```

File: include/deal.II/grid/tria_accessor.templates.h [hunk 2]
```diff
@@ -3173,6 +3173,22 @@ CellAccessor<dim, spacedim>::face(const unsigned int i) const
...
+CellAccessor<dim, spacedim>::face_iterators() const
+{
+  std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>, dim * 2>
```
Comment (masterleinad) [inline include/deal.II/grid/tria_accessor.templates.h]:
```suggestion
  std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>, GeometryInfo<dim>::faces_per_cell>
```

File: include/deal.II/grid/tria_accessor.templates.h [hunk 3]
```diff
@@ -3173,6 +3173,22 @@ CellAccessor<dim, spacedim>::face(const unsigned int i) const
...
+    face_iterators;
+
+  for (unsigned int i = 0; i < dim * 2; i++)
```
Comment (masterleinad) [inline include/deal.II/grid/tria_accessor.templates.h]:
```suggestion
  for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; i++)
```

**Comment by drwells** [comment]
Very nice. We should have added this years ago!

File: include/deal.II/grid/tria_accessor.h [hunk 4]
```diff
@@ -2733,6 +2733,9 @@ class CellAccessor : public TriaAccessor<dim, dim, spacedim>
   TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>
   face(const unsigned int i) const;
 
+  std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>, 2 * dim>
```
Comment (masterleinad) [inline include/deal.II/grid/tria_accessor.h]:
```suggestion
  std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>, GeometryInfo<dim>::faces_per_cell>
```

**Comment by bangerth** [comment]
On 8/14/19 8:00 AM, David Wells wrote:
> We should have added this years ago!

Amen. How did that not ever occur to any of us?

**Comment by peterrum** [comment]
@masterleinad: Thanks for the review! I have addend the documentation, the test, and the change-log entry (I have added @bangerth  and @masterleinad as co-authors, since you both came up with the final concept.).


@bangerth: I think the new test tests/grid/accessor_02.cc looks exactly as you  described it in  #8563.

File: include/deal.II/grid/tria_accessor.templates.h [hunk 5]
```diff
@@ -3173,6 +3173,24 @@ CellAccessor<dim, spacedim>::face(const unsigned int i) const
...
+template <int dim, int spacedim>
+inline std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>,
+                  GeometryInfo<dim>::faces_per_cell>
+CellAccessor<dim, spacedim>::face_iterators() const
+{
+  std::array<TriaIterator<TriaAccessor<dim - 1, dim, spacedim>>,
+             GeometryInfo<dim>::faces_per_cell>
+    face_iterators;
+
+  for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; i++)
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.templates.h]:
```suggestion
  for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; ++i)
```

File: tests/grid/accessor_02.cc [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+template <int dim, int spacedim>
+void
+test()
+{
+  Triangulation<dim, spacedim> tria;
+  GridGenerator::hyper_cube(tria);
+  tria.refine_global(1);
+
+  std::map<typename Triangulation<dim, spacedim>::face_iterator, double> mymap;
```
Comment (bangerth) [inline tests/grid/accessor_02.cc]:
```suggestion
  std::map<typename Triangulation<dim, spacedim>::face_iterator, int> mymap;
```

The cell index is an integer. You'll get a different test output, though.

Comment (peterrum) [reply tests/grid/accessor_02.cc]: I only copied the test `tests/grid/accessor_01.cc`; I did not modify anything except the iterator related stuff. Should I also switch from double to in in this case?

Comment (masterleinad) [reply tests/grid/accessor_02.cc]: Good idea.

Comment (bangerth) [reply tests/grid/accessor_02.cc]: But separate pull request :-)

Comment (peterrum) [reply tests/grid/accessor_02.cc]: @bangerth I have made your requested changes. I will setup the separate PR right away!

**Pull request merged**